crates/sui-framework/packages/sui-framework/sources/crypto/group_ops.move

```
@@ -0,0 +1,120 @@
+
+    // General wrapper for all group elements.
+    struct Element<phantom T> has store, copy, drop {
+        bytes: vector<u8>,
+    }
+
+    public fun bytes<G>(e: &Element<G>): &vector<u8> {
+        &e.bytes
+    }
+    
+    public fun equal<G>(e1: &Element<G>, e2: &Element<G>): bool {
+        &e1.bytes == &e2.bytes
+    }
+
+    // Fails if the bytes are not a valid group element and 'is_trusted' is false.
+    public(friend) fun from_bytes<G>(type: u8, bytes: &vector<u8>, is_trusted: bool): Element<G> {
+        assert!(is_trusted || internal_validate(type, bytes), EInvalidInput);
+        Element<G> { bytes: *bytes }
+    }
+
+    public(friend) fun add<G>(type: u8, e1: &Element<G>, e2: &Element<G>): Element<G> {
+        Element<G> { bytes: internal_add(type, &e1.bytes, &e2.bytes) }
+    }
+
+    public(friend) fun sub<G>(type: u8, e1: &Element<G>, e2: &Element<G>): Element<G> {
+        Element<G> { bytes: internal_sub(type, &e1.bytes, &e2.bytes) }
+    }
+
+    public(friend) fun mul<S, G>(type: u8, scalar: &Element<S>, e: &Element<G>): Element<G> {
+        Element<G> { bytes: internal_mul(type, &scalar.bytes, &e.bytes) }
+    }
+
+    /// Fails if scalar = 0. Else returns 1/scalar * e.
+    public(friend) fun div<S, G>(type: u8, scalar: &Element<S>, e: &Element<G>): Element<G> {
+        Element<G> { bytes: internal_div(type, &scalar.bytes, &e.bytes) }
+    }
+
+    public(friend) fun hash_to<G>(type: u8, m: &vector<u8>): Element<G> {
+        Element<G> { bytes: internal_hash_to(type, m) }
+    }
+
+    /// Aborts with `EInputTooLong` if the vectors are too long.
+    public(friend) fun multi_scalar_multiplication<S, G>(type: u8, scalars: &vector<Element<S>>, elements: &vector<Element<G>>): Element<G> {
+        assert!(vector::length(scalars) == vector::length(elements), EInvalidInput);
+        assert!(vector::length(scalars) > 0, EInvalidInput);
```

I'd swap these 2 asserts for fail fast on empty vector.